iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageInstallationScriptExecutionCheck.java

```diff
@@ -0,0 +1,155 @@
+  }
+
+  private static void checkRunInstruction(CheckContext ctx, RunInstruction instruction) {
+    StatementValidator validator = new StatementValidator(ctx, instruction.arguments());
+    validator.processArguments();
+  }
+
+  private static class StatementValidator {
+
+    CheckContext ctx;
+    List<Argument> arguments;
+    PackageManagerStatement currentStatement;
+
+    StatementValidator(CheckContext ctx, List<Argument> arguments) {
+      this.ctx = ctx;
+      this.arguments = arguments;
+    }
+
+    void processArguments() {
+      if (arguments == null || arguments.isEmpty()) {
+        return;
+      }
+      arguments.forEach(this::process);
+      reportSensitiveStatement();
+    }
+
+    void process(Argument argument) {
+      ArgumentResolution resolution = ArgumentResolution.of(argument);
+
+      // stop analyzing package-manager statement when unresolved part is detected
+      if (resolution.is(UNRESOLVED)) {
+        currentStatement = null;
+        return;
+      }
+      String argValue = resolution.value();
+
+      // if new package-manager statement starts report existing one and create new statement
+      if (TRIGGER_COMMANDS_STATEMENT_TYPE_MAPPING.containsKey(argValue)) {
+        reportSensitiveStatement();
+        currentStatement = new PackageManagerStatement(argument, REQUIRED_FLAG, TRIGGER_COMMANDS_STATEMENT_TYPE_MAPPING.get(argValue));
+      } else if (currentStatement != null) {
+        // check for required flag or store irrelevant ones to the statement
+        if (argValue.startsWith("-")) {
+          if (currentStatement.requiredFlag.equals(argValue)) {
+            currentStatement = null;
+          } else {
+            currentStatement.add(argument);
+          }
+          // detect if statement is installed command or known alias
+        } else if ("install".equals(argValue) || (StatementType.NPM == currentStatement.statementType && NPM_INSTALL_ALIASES.contains(argValue))) {
+          currentStatement.isInstallStatement = true;
+          currentStatement.add(argument);
+          // report statement if install statement without required flag
+        } else {
+          reportSensitiveStatement();
+        }
+      }
+
+    }
+
+    private void reportSensitiveStatement() {
+      if (currentStatement != null && (currentStatement.isInstallStatement ||
+        (StatementType.YARN == currentStatement.statementType && arguments.size() == 1))) {
+        // also reports the statement if the statement consists only of the "yarn" command
+        ctx.reportIssue(mergeArgumentTextRange(currentStatement.arguments), MESSAGE);
+        currentStatement = null;
+      }
+    }
+
+    private static TextRange mergeArgumentTextRange(List<Argument> arguments) {
+      List<TextRange> textRanges = new ArrayList<>();
+      for (Argument argument : arguments) {
+        textRanges.add(argument.textRange());
+      }
+      return TextRanges.merge(textRanges);
+    }
```

Comment:
Let's move this logic to `TextRanges.merge(List<HasTextRange> elementsWithTextRange)`. This way you can replace these lines also in `org.sonar.iac.docker.checks.PackageInstallationCheck`.